### PR TITLE
Chained request ids

### PR DESCRIPTION
**Pull request by rustyrussell**
This turned out to be harder than expected, but this puts the infrastructure in place to create JSON ids which have useful meanings for debugging, eg:

```
pytest:withdraw#11/cln:withdraw#34/txprepare:sendpsbt#4/cln:sendrawtransaction#35
```

This is pytest calling withdraw, which triggers cln to call withdraw, which causes txprepare to call sendpsbt, which causes cln to call sendrawtransaction.

However, cln-rpc needs to be reworked to do this!  Help @cdecker!

Changelog-None

### CODE REVIEW

**Review by vincenzopalazzo — COMMENTED** [review]
Concept ACK, but I partial review it

File: cln-rpc/src/jsonrpc.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use std::fmt::Debug;
 
 #[derive(Debug)]
 pub enum JsonRpc<N, R> {
-    Request(usize, R),
+    Request(String, R),
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply cln-rpc/src/jsonrpc.rs]: I ended up using the serde_json::Value as @cdecker requested.

Still need the cln-rpc to support the correct id form and chaining them based on incoming request though!

**Comment by rustyrussell** [comment]
Rebased on master, removed final commit which had some random fd debugging mixed in!

**Comment by rustyrussell** [comment]
Ha!  Failure was because I didn't port contrib/plugins/cowsay.sh to allow non-numeric IDs.  Fair enough, I'd forgotten it existed...

**Comment by niftynei** [comment]
Can't wait to start using this to track down things! Awesome upgrade. Few minor nits on formatting of strings+spelling, but otherwise, lgtm.

shellcheck is mad at cowsay now..

```
In contrib/plugins/cowsay.sh line 22:
echo '{"jsonrpc":"2.0","id":'$id',"result":{"dynamic":true,"options":[],"rpcmethods":[{"name":"cowsay","usage":"<string>","description":"Have a cow, man!"}]}}'
                             ^-^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
echo '{"jsonrpc":"2.0","id":'"$id"',"result":{"dynamic":true,"options":[],"rpcmethods":[{"name":"cowsay","usage":"<string>","description":"Have a cow, man!"}]}}'
```

**Review by niftynei — APPROVED** [review]

File: lightningd/jsonrpc.c [hunk 2]
```diff
@@ -909,6 +909,8 @@ parse_request(struct json_connection *jcon, const jsmntok_t tok[])
 				    "Expected string for method");
 	}
 
+	log_debug(jcon->log, "IN:id=%s", c->id);
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/jsonrpc.c:915]: Yeah, it's pretty chatty!  But IO level already has the complete JSON.   Let me see what I can do here...

**Pull request merged**